examples/mnist_dataset_api.py

```diff
@@ -0,0 +1,113 @@
+A Dataset is a sequence of elements, which are themselves
+composed of tf.Tensor components. For more details, see:
+https://www.tensorflow.org/programmers_guide/datasets
+
+To use this with Keras, we make a dataset out of elements
+of the form (input batch, output batch). From there, we
+create a one-shot iterator and a graph node corresponding
+to its get_next() method. Its components are then provided
+to the network's Input layer and the Model.compile() method,
+respectively.
+
+Note that from TensorFlow 1.4, tf.contrib.data is deprecated
+and tf.data is preferred. See the release notes for details.
+
+This example is intended to closely follow the
+mnist_tfrecord.py example.
+'''
+from __future__ import division
+
+import numpy as np
+import os
+import tempfile
+
+import keras
+from keras import backend as K
+from keras import layers
+from keras.datasets import mnist
+
+import tensorflow as tf
+from tensorflow.contrib.data import Dataset
+
+
+if K.backend() != 'tensorflow':
+    raise RuntimeError('This example can only run with the TensorFlow backend,'
+                       ' because it requires the Datset API, which is not'
+                       ' supported on other platforms.')
+
+
+def cnn_layers(inputs):
+    x = layers.Conv2D(32, (3, 3),
+                      activation='relu', padding='valid')(inputs)
+    x = layers.MaxPooling2D(pool_size=(2, 2))(x)
+    x = layers.Conv2D(64, (3, 3), activation='relu')(x)
+    x = layers.MaxPooling2D(pool_size=(2, 2))(x)
+    x = layers.Flatten()(x)
+    x = layers.Dense(512, activation='relu')(x)
+    x = layers.Dropout(0.5)(x)
+    predictions = layers.Dense(classes,
+                               activation='softmax',
+                               name='x_train_out')(x)
+    return predictions
+
+
+batch_size = 128
+buffer_size = 10000
+steps_per_epoch = np.ceil(60000 / 128).astype('int')  # = 469
+epochs = 5
+classes = 10
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+x_train = x_train.astype(np.float32) / 255
+x_train = np.expand_dims(x_train, -1)
+y_train = tf.one_hot(y_train, classes)
+
+# Create the dataset and its associated one-shot iterator.
+dataset = Dataset.from_tensor_slices((x_train, y_train))
+dataset = dataset.repeat()
+dataset = dataset.shuffle(buffer_size)
+dataset = dataset.batch(batch_size)
+iterator = dataset.make_one_shot_iterator()
+
+# Model creation using tensors from the get_next() graph node.
+inputs, targets = iterator.get_next()
+model_input = layers.Input(tensor=inputs)
+model_output = cnn_layers(model_input)
+train_model = keras.models.Model(inputs=model_input, outputs=model_output)
+
+train_model.compile(optimizer=keras.optimizers.RMSprop(lr=2e-3, decay=1e-5),
+                    loss='categorical_crossentropy',
+                    metrics=['accuracy'],
+                    target_tensors=[targets])
+train_model.summary()
+
+train_model.fit(epochs=epochs,
+                steps_per_epoch=steps_per_epoch)
```

Comment:
As a user I'm wondering if it is possible to use validation data together with this workflow?

Reply:
On top of my mind, I think you can, but the model from draw validation samples/targets from its hard-coded queue, which in this example is the training data. Would have to check though.

Reply:
@fchollet Thanks for the fast answer.

For the record, I just naively tried adding `validation_data=(x_val, y_val)` to the `.fit` call, but it errors:
```bash
File "keras/engine/training.py", line 59, in _standardize_input_data
  'expected no data, but got:', data)
ValueError: ('Error when checking model input: expected no data, but got:', array([[[[ 0.],
...
```

Also using `validation_split` fails, but I guess that's expected:
```bash
File "keras/engine/training.py", line 1584, in fit
    if hasattr(x[0], 'shape'):
IndexError: list index out of range
```